### PR TITLE
Ublox SARA-N310 A2. Added PSM.

### PR DESCRIPTION
**Pull request by Jeff-H2**
Added PSM functionality on top of basic socket connectivity. The user can configure PSM at runtime. The V_INT pin is used to check if the module is asleep.

Keywords: NVSETPM, CPSMS, CSCLK, NSOSTF, modem pins.

### CODE REVIEW

File: drivers/modem/ublox-sara-n310.c [hunk 1]
```diff
@@ -155,6 +164,52 @@ static int modem_atoi(const char *s, const int err_value, const char *desc,
...
+		ret = modem_cmd_send_nolock(iface, handler, handler_cmds,
+					    handler_cmds_len, buf, sem,
+					    timeout);
+	}
+
+	return ret;
+}
+
+/* network state getter */
+int n310_get_state(void)
```
Comment (abelNoelleC) [inline drivers/modem/ublox-sara-n310.c]: maybe change function name to something more explicit (what state is it getting?)

File: drivers/modem/ublox-sara-n310.c [hunk 2]
```diff
@@ -491,9 +553,9 @@ static int create_socket(struct modem_socket *sock, const struct sockaddr *addr)
 		snprintk(buf, sizeof(buf), "AT+USOCR=%d", proto);
 	}
 
-	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
-			     &cmd, 1U, buf, &mdata.sem_response,
-			     MDM_CMD_TIMEOUT);
+	ret = send_at_command(&mdata.context.iface, &mdata.context.cmd_handler,
```
Comment (abelNoelleC) [inline drivers/modem/ublox-sara-n310.c:553]: why did you change the function? are there specific pros/cons for the change, and/or consequences?

File: samples/drivers/ublox_sara_n310/src/main.c [hunk 3]
```diff
@@ -36,46 +41,48 @@ void main(void)
...
+		LOG_INF("Sending...");
 		n = sendto(sockfd, buffer, strlen(buffer) / 2,
 			   RELEASE_AFTER_FIRST_DOWNLINK,
 			   (const struct sockaddr *)&addr, sizeof(addr));
 		if (n < 0) {
 			LOG_ERR("Send failed with return value: %d", n);
 		}
 
-		k_msleep(2000);
+		k_msleep(5000);
```
Comment (abelNoelleC) [inline samples/drivers/ublox_sara_n310/src/main.c:76]: is there a specific reason for these sleep lengths?

**Review by abelNoelleC — COMMENTED** [review]
Looks good, added some small comments/questions

File: drivers/modem/ublox-sara-n310.c (hunk 2 above)
Comment (Jeff-H2) [reply drivers/modem/ublox-sara-n310.c:553]: > why did you change the function? are there specific pros/cons for the change, and/or consequences?

This was because I've paired checking if the module is awake and sending an AT command into the send_at_command function. It saved me from having to repeat myself (check if module is awake) every time I wanted to use modem_cmd_send.

File: samples/drivers/ublox_sara_n310/src/main.c (hunk 3 above)
Comment (Jeff-H2) [reply samples/drivers/ublox_sara_n310/src/main.c:76]: No profound reason, I was just checking different delays I think. I also worried a little if I was using too much bandwidth while testing.

File: drivers/modem/ublox-sara-n310.c [hunk 4]
```diff
@@ -256,6 +311,19 @@ MODEM_CMD_DEFINE(on_cmd_atcmdinfo_model)
 	return 0;
 }
 
+/* CMD: +NPSMR */
+MODEM_CMD_DEFINE(on_cmd_npsmr)
+{
+	size_t out_len;
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c]: Make a good practice to always initialize your variables

File: drivers/modem/ublox-sara-n310.c [hunk 5]
```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
...
+ * Pin functions.
+ *
+ * The SARA-N310 forces the power pin low to achieve the following:
+ * min 1s - max 2.5s: trigger module switch-on
+ * min 1s - max 2.5s: trigger module wake-up from PSM
+ * min 2.5s: trigger module switch-off
+ */
+static int turn_on_module(void)
+{
+	int ret, timeout = 0;
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c]: You know that here only one variable is initialized ,
If you want to do it in one line
```
int ret=0, timeout = 0;
```

I am always in favor though of declaring vars in separate lines.

File: drivers/modem/ublox-sara-n310.c [hunk 6]
```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
...
 
+/*
+ * Pin functions.
+ *
+ * The SARA-N310 forces the power pin low to achieve the following:
+ * min 1s - max 2.5s: trigger module switch-on
+ * min 1s - max 2.5s: trigger module wake-up from PSM
+ * min 2.5s: trigger module switch-off
+ */
+static int turn_on_module(void)
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c:909]: Could you a add a small description what this function returns ?

File: drivers/modem/ublox-sara-n310.c [hunk 7]
```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
...
+	k_sleep(K_MSEC(1500));
+
+	LOG_DBG("MDM_POWER_PIN -> DISABLE");
+	modem_pin_write(&mdata.context, MDM_POWER, MDM_POWER_DISABLE);
+
+	while (!is_awake() && (timeout++ < 5)) {
+		k_sleep(K_MSEC(10));
+	}
+
+	if (timeout >= 5) {
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c:925]: Please add a comment why is this needed?

File: drivers/modem/ublox-sara-n310.c [hunk 8]
```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
...
+
+/* check if module is awake by checking V_INT value */
+static int is_awake(void)
+{
+	return modem_pin_read(&mdata.context, MDM_VINT);
+}
+
+/* PSM functions, exposed to application through header */
+/* Power mode setting, +NVSETPM */
+int n310_psm_set_mode(int psm_mode)
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c:954]: [Minor and optional] Would be nice to have an enum with the allowed states. In this way you can help the user of the function

File: drivers/modem/ublox-sara-n310.c [hunk 9]
```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
...
+{
+	int ret;
+	char buf[sizeof("AT+NVSETPM=##\r")];
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c]: Always initialize your variables
```
int ret=0;
char buf[sizeof("AT+NVSETPM=##\r")] ={0};
```

File: drivers/modem/ublox-sara-n310.c [hunk 10]
```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
...
+	}
+
+	return ret;
+}
+
+/* Low clock mode setting, +CSCLK */
+int n310_psm_set_csclk(int setting)
+{
+	int ret;
+	char buf[sizeof("AT+CSCLK=#\r")];
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c]: Same here about initialization

File: drivers/modem/ublox-sara-n310.c [hunk 11]
```diff
@@ -835,6 +897,148 @@ static bool offload_is_supported(int family, int type, int proto)
...
+	}
+
+	return ret;
+}
+
+/* Power Saving Mode setting, +CPSMS */
+int n310_psm_config(int mode, char *periodic_TAU, char *active_time)
+{
+	int ret;
+	char buf[sizeof("AT+CPSMS=#,,,\"########\",\"########\"\r")];
```
Comment (abelkarapatis) [inline drivers/modem/ublox-sara-n310.c]: I guess you know what I am going to say :)

File: include/drivers/modem/ublox-sara-n310.h [hunk 12]
```diff
@@ -15,6 +15,23 @@
 #define RELEASE_AFTER_UPLINK 1
 #define RELEASE_AFTER_FIRST_DOWNLINK 2
 
+/* PSM defines */
+#define PM0 0
```
Comment (abelkarapatis) [inline include/drivers/modem/ublox-sara-n310.h:19]: Nice ones, why not in a  enum?

**Review by abelkarapatis — COMMENTED** [review]
Review done, some small remarks on my side and I guess we can merge it
